File: tools/reference/src/math_probe.cpp
#include "math_probe.hpp"

#include "protocol.hpp"

#include <Box2D/Common/b2Math.h>

#include <stdexcept>

namespace liquidfun::reference {
namespace {

b2Vec2 vec2(const Vec2Bits& bits) {
  return {float_from_bits(bits.x), float_from_bits(bits.y)};
}

b2Vec3 vec3(const Vec3Bits& bits) {
  return {float_from_bits(bits.x), float_from_bits(bits.y),
          float_from_bits(bits.z)};
}

b2Mat22 mat22(const Mat22Bits& bits) {
  return {vec2(bits.first), vec2(bits.second)};
}

b2Mat33 mat33(const Mat33Bits& bits) {
  return {vec3(bits.first), vec3(bits.second), vec3(bits.third)};
}

b2Transform transform(const TransformBits& bits) {
  return {vec2(bits.position), b2Rot(float_from_bits(bits.angle))};
}

b2Sweep sweep(const SweepBits& bits) {
  b2Sweep value;
  value.localCenter = vec2(bits.local_center);
  value.c0 = vec2(bits.initial_center);
  value.c = vec2(bits.center);
  value.a0 = float_from_bits(bits.initial_angle);
  value.a = float_from_bits(bits.angle);
  value.alpha0 = float_from_bits(bits.initial_fraction);
  return value;
}

void push(
    std::vector<MathProbeValue>& values,
    MathProbeValueField field,
    float value) {
  values.push_back({field, bits_from_float(value)});
}

void push_vec2(std::vector<MathProbeValue>& values, const b2Vec2& value) {
  push(values, MathProbeValueField::x, value.x);
  push(values, MathProbeValueField::y, value.y);
}

void push_position(
    std::vector<MathProbeValue>& values,
    const b2Vec2& value) {
  push(values, MathProbeValueField::position_x, value.x);
  push(values, MathProbeValueField::position_y, value.y);
}

void push_vec3(std::vector<MathProbeValue>& values, const b2Vec3& value) {
  push(values, MathProbeValueField::x, value.x);
  push(values, MathProbeValueField::y, value.y);
  push(values, MathProbeValueField::z, value.z);
}

void push_mat22(std::vector<MathProbeValue>& values, const b2Mat22& value) {
  push_vec2(values, value.ex);
  push_vec2(values, value.ey);
}

void push_mat33(std::vector<MathProbeValue>& values, const b2Mat33& value) {
  push_vec3(values, value.ex);
  push_vec3(values, value.ey);
  push_vec3(values, value.ez);
}

MathProbeResult execute_case(const MathProbeCase& probe) {
  MathProbeResult result{
      probe.case_id,
      probe.operation,
      probe.policy_path,
      probe.horizon,
      {},
      {}};
  switch (probe.operation) {
    case MathProbeOperation::is_valid: {
      const auto& input = std::get<ScalarInput>(probe.input);
      result.discrete.push_back({
          MathProbeDiscreteField::predicate,
          b2IsValid(float_from_bits(input.value))});
      break;
    }
    case MathProbeOperation::abs: {
      const auto& input = std::get<ScalarInput>(probe.input);
      push(result.values, MathProbeValueField::value,
           b2Abs(float_from_bits(input.value)));
      break;
    }
    case MathProbeOperation::min:
    case MathProbeOperation::max: {
      const auto& input = std::get<BinaryInput>(probe.input);
      const auto a = float_from_bits(input.a);
      const auto b = float_from_bits(input.b);
      push(result.values, MathProbeValueField::value,
           probe.operation == MathProbeOperation::min ? b2Min(a, b)
                                                      : b2Max(a, b));
      break;
    }
    case MathProbeOperation::clamp: {
      const auto& input = std::get<ClampInput>(probe.input);
      push(result.values, MathProbeValueField::value,
           b2Clamp(float_from_bits(input.value), float_from_bits(input.low),
                   float_from_bits(input.high)));
      break;
    }
    case MathProbeOperation::inv_sqrt: {
      const auto& input = std::get<ScalarInput>(probe.input);
      push(result.values, MathProbeValueField::value,
           b2InvSqrt(float_from_bits(input.value)));
      break;
    }
    case MathProbeOperation::vec_length: {
      const auto& input = std::get<Vector2Input>(probe.input);
      push(result.values, MathProbeValueField::length,
           vec2(input.vector).Length());
      break;
    }
    case MathProbeOperation::vec_normalize: {
      const auto& input = std::get<Vector2Input>(probe.input);
      auto value = vec2(input.vector);
      const auto length = value.Normalize();
      push(result.values, MathProbeValueField::length, length);
      push_vec2(result.values, value);
      result.discrete.push_back(
          {MathProbeDiscreteField::normalized, length != 0.0F});
      break;
    }
    case MathProbeOperation::dot:
    case MathProbeOperation::cross: {
      const auto& input = std::get<VectorPairInput>(probe.input);
      const auto a = vec2(input.a);
      const auto b = vec2(input.b);
      push(result.values, MathProbeValueField::value,
           probe.operation == MathProbeOperation::dot ? b2Dot(a, b)
                                                      : b2Cross(a, b));
      break;
    }
    case MathProbeOperation::mat22_solve: {
      const auto& input = std::get<Mat22SolveInput>(probe.input);
      const auto matrix = mat22(input.matrix);
      const auto determinant =
          matrix.ex.x * matrix.ey.y - matrix.ey.x * matrix.ex.y;
      push_vec2(result.values, matrix.Solve(vec2(input.right)));
      result.discrete.push_back(
          {MathProbeDiscreteField::non_zero_determinant,
           determinant != 0.0F});
      break;
    }
    case MathProbeOperation::mat33_solve: {
      const auto& input = std::get<Mat33SolveInput>(probe.input);
      const auto matrix = mat33(input.matrix);
      const auto determinant = b2Dot(matrix.ex, b2Cross(matrix.ey, matrix.ez));
      push_vec3(result.values, matrix.Solve33(vec3(input.right)));
      result.discrete.push_back(
          {MathProbeDiscreteField::non_zero_determinant,
           determinant != 0.0F});
      break;
    }
    case MathProbeOperation::mat22_inverse: {
      const auto& input = std::get<Mat22Input>(probe.input);
      const auto matrix = mat22(input.matrix);
      const auto determinant =
          matrix.ex.x * matrix.ey.y - matrix.ey.x * matrix.ex.y;
      push_mat22(result.values, matrix.GetInverse());
      result.discrete.push_back(
          {MathProbeDiscreteField::non_zero_determinant,
           determinant != 0.0F});
      break;
    }
    case MathProbeOperation::mat33_sym_inverse: {
      const auto& input = std::get<Mat33Input>(probe.input);
      const auto matrix = mat33(input.matrix);
      const auto determinant = b2Dot(matrix.ex, b2Cross(matrix.ey, matrix.ez));
      b2Mat33 inverse;
      matrix.GetSymInverse33(&inverse);
      push_mat33(result.values, inverse);
      result.discrete.push_back(
          {MathProbeDiscreteField::non_zero_determinant,
           determinant != 0.0F});
      break;
    }
    case MathProbeOperation::rotation: {
      const auto& input = std::get<RotationInput>(probe.input);
      const b2Rot value(float_from_bits(input.angle));
      push(result.values, MathProbeValueField::sine, value.s);
      push(result.values, MathProbeValueField::cosine, value.c);
      push(result.values, MathProbeValueField::angle, value.GetAngle());
      break;
    }
    case MathProbeOperation::transform: {
      const auto& input = std::get<TransformInput>(probe.input);
      auto composed = transform(input.left);
      for (std::uint32_t index = 0; index < probe.horizon.steps; ++index) {
        composed = b2Mul(composed, transform(input.right));
      }
      push_position(result.values, b2Mul(composed, vec2(input.point)));
      push(result.values, MathProbeValueField::angle, composed.q.GetAngle());
      break;
    }
    case MathProbeOperation::sweep_transform: {
      const auto& input = std::get<SweepTransformInput>(probe.input);
      b2Transform value;
      sweep(input.sweep).GetTransform(&value, float_from_bits(input.fraction));
      push_position(result.values, value.p);
      push(result.values, MathProbeValueField::sine, value.q.s);
      push(result.values, MathProbeValueField::cosine, value.q.c);
      break;
    }
    case MathProbeOperation::sweep_advance: {
      const auto& input = std::get<SweepAdvanceInput>(probe.input);
      auto value = sweep(input.sweep);
      for (const auto fraction : input.fractions) {
        value.Advance(float_from_bits(fraction));
      }
      push(result.values, MathProbeValueField::initial_center_x, value.c0.x);
      push(result.values, MathProbeValueField::initial_center_y, value.c0.y);
      push(result.values, MathProbeValueField::initial_angle, value.a0);
      push(result.values, MathProbeValueField::initial_fraction, value.alpha0);
      result.discrete.push_back({MathProbeDiscreteField::advanced, true});
      break;
    }
    case MathProbeOperation::sweep_normalize: {
      const auto& input = std::get<SweepInput>(probe.input);
      auto value = sweep(input.sweep);
      value.Normalize();
      push(result.values, MathProbeValueField::initial_angle, value.a0);
      push(result.values, MathProbeValueField::angle, value.a);
      break;
    }
    case MathProbeOperation::cancellation: {
      const auto& input = std::get<CancellationInput>(probe.input);
      const auto large = float_from_bits(input.large);
      const auto opposite = float_from_bits(input.opposite);
      const auto tail = float_from_bits(input.tail);
      const auto left_pair = large + opposite;
      const auto right_pair = opposite + tail;
      push(result.values, MathProbeValueField::left_associated,
           left_pair + tail);
      push(result.values, MathProbeValueField::right_associated,
           large + right_pair);
      break;
    }
    case MathProbeOperation::halfway_rounding: {
      const auto& input = std::get<HalfwayRoundingInput>(probe.input);
      push(result.values, MathProbeValueField::even_midpoint,
           float_from_bits(input.even) + float_from_bits(input.half_ulp));
      push(result.values, MathProbeValueField::odd_midpoint,
           float_from_bits(input.odd) + float_from_bits(input.half_ulp));
      break;
    }
    case MathProbeOperation::overflow:
    case MathProbeOperation::underflow: {
      const auto& input = std::get<ScaleInput>(probe.input);
      push(result.values, MathProbeValueField::value,
           float_from_bits(input.value) * float_from_bits(input.factor));
      break;
    }
    case MathProbeOperation::fma_witness: {
      const auto& input = std::get<FmaWitnessInput>(probe.input);
      const auto product = float_from_bits(input.a) * float_from_bits(input.b);
      push(result.values, MathProbeValueField::value,
           product + float_from_bits(input.c));
      break;
    }
  }
  return result;
}

}  // namespace

std::vector<MathProbeResult> execute_math_probe(
    const MathProbeRequest& request) {
  std::vector<MathProbeResult> results;
  results.reserve(request.cases.size());
  for (const auto& probe : request.cases) {
    results.push_back(execute_case(probe));
  }
  return results;
}

}  // namespace liquidfun::reference

File: crates/liquidfun-differential/src/math_probe.rs
//! Pure, bounded native execution for the Phase 4 math-probe contract.

use liquidfun::math::{self, Mat22, Mat33, Rotation, Sweep, SweepError, Transform, Vec2, Vec3};
use liquidfun_test_protocol::{
    FloatBits, Mat22Bits, Mat33Bits, MathProbeCase, MathProbeDiscrete, MathProbeDiscreteField,
    MathProbeInput, MathProbeOperation, MathProbeRequestRecord, MathProbeResult, MathProbeValue,
    MathProbeValueField, SweepBits, TransformBits, Vec2Bits, Vec3Bits,
};

/// Failure constructing or advancing a checked sweep from validated probe bits.
#[derive(Debug, thiserror::Error)]
pub enum MathProbeExecutionError {
    /// A structured sweep violates the checked public math invariant.
    #[error("case {case_id} contains an invalid checked sweep: {source}")]
    Sweep {
        /// Stable case identity.
        case_id: Box<str>,
        #[source]
        /// Typed checked-sweep failure.
        source: SweepError,
    },
}

/// Stateless native executor; it creates no world, solver, or persistent adapter state.
pub struct NativeMathProbeExecutor;

impl NativeMathProbeExecutor {
    /// Executes all validated cases in request order.
    pub fn execute(
        request: &MathProbeRequestRecord,
    ) -> Result<Box<[MathProbeResult]>, MathProbeExecutionError> {
        request
            .scenario()
            .cases()
            .iter()
            .map(execute_case)
            .collect::<Result<Vec<_>, _>>()
            .map(Vec::into_boxed_slice)
    }
}

fn execute_case(case: &MathProbeCase) -> Result<MathProbeResult, MathProbeExecutionError> {
    let mut values = Vec::new();
    let mut discrete = Vec::new();
    match (case.operation(), case.input()) {
        (MathProbeOperation::IsValid, MathProbeInput::Scalar { value_bits }) => {
            discrete.push(discrete_value(
                MathProbeDiscreteField::Predicate,
                math::is_valid(value_bits.to_f32()),
            ));
        }
        (MathProbeOperation::Abs, MathProbeInput::Scalar { value_bits }) => {
            push(
                &mut values,
                MathProbeValueField::Value,
                math::abs(value_bits.to_f32()),
            );
        }
        (
            operation @ (MathProbeOperation::Min | MathProbeOperation::Max),
            MathProbeInput::Binary { a_bits, b_bits },
        ) => {
            let value = if operation == MathProbeOperation::Min {
                math::min(a_bits.to_f32(), b_bits.to_f32())
            } else {
                math::max(a_bits.to_f32(), b_bits.to_f32())
            };
            push(&mut values, MathProbeValueField::Value, value);
        }
        (
            MathProbeOperation::Clamp,
            MathProbeInput::Clamp {
                value_bits,
                low_bits,
                high_bits,
            },
        ) => {
            push(
                &mut values,
                MathProbeValueField::Value,
                math::clamp(value_bits.to_f32(), low_bits.to_f32(), high_bits.to_f32()),
            );
        }
        (MathProbeOperation::InvSqrt, MathProbeInput::Scalar { value_bits }) => {
            push(
                &mut values,
                MathProbeValueField::Value,
                math::inverse_sqrt(value_bits.to_f32()),
            );
        }
        (MathProbeOperation::VecLength, MathProbeInput::Vector2 { vector }) => {
            push(
                &mut values,
                MathProbeValueField::Length,
                vec2(*vector).length(),
            );
        }
        (MathProbeOperation::VecNormalize, MathProbeInput::Vector2 { vector }) => {
            let mut vector = vec2(*vector);
            let length = vector.normalize();
            push(&mut values, MathProbeValueField::Length, length);
            push_vec2(&mut values, vector);
            discrete.push(discrete_value(
                MathProbeDiscreteField::Normalized,
                length != 0.0,
            ));
        }
        (
            operation @ (MathProbeOperation::Dot | MathProbeOperation::Cross),
            MathProbeInput::VectorPair { a, b },
        ) => {
            let a = vec2(*a);
            let b = vec2(*b);
            let value = if operation == MathProbeOperation::Dot {
                a.dot(b)
            } else {
                a.cross(b)
            };
            push(&mut values, MathProbeValueField::Value, value);
        }
        (MathProbeOperation::Mat22Solve, MathProbeInput::Mat22Solve { matrix, right }) => {
            let determinant = matrix.first.x_bits.to_f32() * matrix.second.y_bits.to_f32()
                - matrix.second.x_bits.to_f32() * matrix.first.y_bits.to_f32();
            push_vec2(&mut values, mat22(*matrix).solve(vec2(*right)));
            discrete.push(discrete_value(
                MathProbeDiscreteField::NonZeroDeterminant,
                determinant != 0.0,
            ));
        }
        (MathProbeOperation::Mat33Solve, MathProbeInput::Mat33Solve { matrix, right }) => {
            let matrix = mat33(*matrix);
            let determinant = matrix
                .first_column()
                .dot(matrix.second_column().cross(matrix.third_column()));
            push_vec3(&mut values, matrix.solve33(vec3(*right)));
            discrete.push(discrete_value(
                MathProbeDiscreteField::NonZeroDeterminant,
                determinant != 0.0,
            ));
        }
        (MathProbeOperation::Mat22Inverse, MathProbeInput::Mat22 { matrix }) => {
            let determinant = matrix.first.x_bits.to_f32() * matrix.second.y_bits.to_f32()
                - matrix.second.x_bits.to_f32() * matrix.first.y_bits.to_f32();
            push_mat22(&mut values, mat22(*matrix).inverse());
            discrete.push(discrete_value(
                MathProbeDiscreteField::NonZeroDeterminant,
                determinant != 0.0,
            ));
        }
        (MathProbeOperation::Mat33SymInverse, MathProbeInput::Mat33 { matrix }) => {
            let matrix = mat33(*matrix);
            let determinant = matrix
                .first_column()
                .dot(matrix.second_column().cross(matrix.third_column()));
            push_mat33(&mut values, matrix.symmetric_inverse33());
            discrete.push(discrete_value(
                MathProbeDiscreteField::NonZeroDeterminant,
                determinant != 0.0,
            ));
        }
        (MathProbeOperation::Rotation, MathProbeInput::Rotation { angle_bits }) => {
            let rotation = Rotation::from_angle(angle_bits.to_f32());
            push(&mut values, MathProbeValueField::Sine, rotation.sine());
            push(&mut values, MathProbeValueField::Cosine, rotation.cosine());
            push(&mut values, MathProbeValueField::Angle, rotation.angle());
        }
        (MathProbeOperation::Transform, MathProbeInput::Transform { left, right, point }) => {
            let mut composed = transform(*left);
            for _ in 0..case.horizon().steps() {
                composed = composed.compose(transform(*right));
            }
            let result = composed.apply(vec2(*point));
            push_position(&mut values, result);
            push(
                &mut values,
                MathProbeValueField::Angle,
                composed.rotation().angle(),
            );
        }
        (
            MathProbeOperation::SweepTransform,
            MathProbeInput::SweepTransform {
                sweep,
                fraction_bits,
            },
        ) => {
            let transform = sweep_from(case, *sweep)?.transform_at(fraction_bits.to_f32());
            push_position(&mut values, transform.position());
            push(
                &mut values,
                MathProbeValueField::Sine,
                transform.rotation().sine(),
            );
            push(
                &mut values,
                MathProbeValueField::Cosine,
                transform.rotation().cosine(),
            );
        }
        (
            MathProbeOperation::SweepAdvance,
            MathProbeInput::SweepAdvance {
                sweep,
                fractions_bits,
            },
        ) => {
            let mut sweep = sweep_from(case, *sweep)?;
            for fraction in fractions_bits.as_slice() {
                sweep
                    .advance(fraction.to_f32())
                    .map_err(|source| sweep_error(case, source))?;
            }
            push_sweep_state(&mut values, sweep);
            discrete.push(discrete_value(MathProbeDiscreteField::Advanced, true));
        }
        (MathProbeOperation::SweepNormalize, MathProbeInput::Sweep { sweep }) => {
            let mut sweep = sweep_from(case, *sweep)?;
            sweep.normalize();
            push(
                &mut values,
                MathProbeValueField::InitialAngle,
                sweep.initial_angle(),
            );
            push(&mut values, MathProbeValueField::Angle, sweep.angle());
        }
        (
            MathProbeOperation::Cancellation,
            MathProbeInput::Cancellation {
                large_bits,
                opposite_bits,
                tail_bits,
            },
        ) => {
            let left_pair = large_bits.to_f32() + opposite_bits.to_f32();
            let left = left_pair + tail_bits.to_f32();
            let right_pair = opposite_bits.to_f32() + tail_bits.to_f32();
            let right = large_bits.to_f32() + right_pair;
            push(&mut values, MathProbeValueField::LeftAssociated, left);
            push(&mut values, MathProbeValueField::RightAssociated, right);
        }
        (
            MathProbeOperation::HalfwayRounding,
            MathProbeInput::HalfwayRounding {
                even_bits,
                odd_bits,
                half_ulp_bits,
            },
        ) => {
            push(
                &mut values,
                MathProbeValueField::EvenMidpoint,
                even_bits.to_f32() + half_ulp_bits.to_f32(),
            );
            push(
                &mut values,
                MathProbeValueField::OddMidpoint,
                odd_bits.to_f32() + half_ulp_bits.to_f32(),
            );
        }
        (
            MathProbeOperation::Overflow | MathProbeOperation::Underflow,
            MathProbeInput::Scale {
                value_bits,
                factor_bits,
            },
        ) => {
            push(
                &mut values,
                MathProbeValueField::Value,
                value_bits.to_f32() * factor_bits.to_f32(),
            );
        }
        (
            MathProbeOperation::FmaWitness,
            MathProbeInput::FmaWitness {
                a_bits,
                b_bits,
                c_bits,
            },
        ) => {
            let product = a_bits.to_f32() * b_bits.to_f32();
            push(
                &mut values,
                MathProbeValueField::Value,
                product + c_bits.to_f32(),
            );
        }
        _ => unreachable!("validated operation/input pairs are exhaustive"),
    }
    Ok(MathProbeResult::new(
        case.case_id(),
        case.operation(),
        case.policy_path(),
        case.horizon(),
        values,
        discrete,
    ))
}

fn vec2(bits: Vec2Bits) -> Vec2 {
    Vec2::new(bits.x_bits.to_f32(), bits.y_bits.to_f32())
}
fn vec3(bits: Vec3Bits) -> Vec3 {
    Vec3::new(
        bits.x_bits.to_f32(),
        bits.y_bits.to_f32(),
        bits.z_bits.to_f32(),
    )
}
fn mat22(bits: Mat22Bits) -> Mat22 {
    Mat22::from_columns(vec2(bits.first), vec2(bits.second))
}
fn mat33(bits: Mat33Bits) -> Mat33 {
    Mat33::from_columns(vec3(bits.first), vec3(bits.second), vec3(bits.third))
}
fn transform(bits: TransformBits) -> Transform {
    Transform::from_position_angle(vec2(bits.position), bits.angle_bits.to_f32())
}

fn sweep_from(case: &MathProbeCase, bits: SweepBits) -> Result<Sweep, MathProbeExecutionError> {
    Sweep::new(
        vec2(bits.local_center),
        vec2(bits.initial_center),
        vec2(bits.center),
        bits.initial_angle_bits.to_f32(),
        bits.angle_bits.to_f32(),
        bits.initial_fraction_bits.to_f32(),
    )
    .map_err(|source| sweep_error(case, source))
}

fn sweep_error(case: &MathProbeCase, source: SweepError) -> MathProbeExecutionError {
    MathProbeExecutionError::Sweep {
        case_id: case.case_id().into(),
        source,
    }
}

fn push(values: &mut Vec<MathProbeValue>, field: MathProbeValueField, value: f32) {
    values.push(MathProbeValue::new(field, FloatBits::from_f32(value)));
}
fn push_vec2(values: &mut Vec<MathProbeValue>, value: Vec2) {
    push(values, MathProbeValueField::X, value.x);
    push(values, MathProbeValueField::Y, value.y);
}
fn push_position(values: &mut Vec<MathProbeValue>, value: Vec2) {
    push(values, MathProbeValueField::PositionX, value.x);
    push(values, MathProbeValueField::PositionY, value.y);
}
fn push_vec3(values: &mut Vec<MathProbeValue>, value: Vec3) {
    push(values, MathProbeValueField::X, value.x);
    push(values, MathProbeValueField::Y, value.y);
    push(values, MathProbeValueField::Z, value.z);
}
fn push_mat22(values: &mut Vec<MathProbeValue>, value: Mat22) {
    push_vec2(values, value.first_column());
    push_vec2(values, value.second_column());
}
fn push_mat33(values: &mut Vec<MathProbeValue>, value: Mat33) {
    push_vec3(values, value.first_column());
    push_vec3(values, value.second_column());
    push_vec3(values, value.third_column());
}
fn push_sweep_state(values: &mut Vec<MathProbeValue>, value: Sweep) {
    push(
        values,
        MathProbeValueField::InitialCenterX,
        value.initial_center().x,
    );
    push(
        values,
        MathProbeValueField::InitialCenterY,
        value.initial_center().y,
    );
    push(
        values,
        MathProbeValueField::InitialAngle,
        value.initial_angle(),
    );
    push(
        values,
        MathProbeValueField::InitialFraction,
        value.initial_fraction(),
    );
}
fn discrete_value(field: MathProbeDiscreteField, value: bool) -> MathProbeDiscrete {
    MathProbeDiscrete::new(field, value)
}

#[cfg(test)]
mod tests {
    use super::*;
    use liquidfun_test_protocol::{HarnessLimits, decode_math_probe_request_jsonl};

    const REQUEST: &[u8] =
        include_bytes!("../../../protocol/fixtures/accepted/math-probe-request.jsonl");

    fn results() -> Box<[MathProbeResult]> {
        let request = decode_math_probe_request_jsonl(REQUEST, &HarnessLimits::phase2_default_v1())
            .expect("checked-in probe request should decode");
        NativeMathProbeExecutor::execute(&request).expect("native probes should execute")
    }

    #[test]
    fn native_math_probe_executes_complete_witness_corpus() {
        // Arrange / Act
        let results = results();

        // Assert
        assert!(results.len() >= 36);
        for operation in [
            MathProbeOperation::IsValid,
            MathProbeOperation::Abs,
            MathProbeOperation::Min,
            MathProbeOperation::Max,
            MathProbeOperation::Clamp,
            MathProbeOperation::InvSqrt,
            MathProbeOperation::VecLength,
            MathProbeOperation::VecNormalize,
            MathProbeOperation::Dot,
            MathProbeOperation::Cross,
            MathProbeOperation::Mat22Solve,
            MathProbeOperation::Mat33Solve,
            MathProbeOperation::Mat22Inverse,
            MathProbeOperation::Mat33SymInverse,
            MathProbeOperation::Rotation,
            MathProbeOperation::Transform,
            MathProbeOperation::SweepTransform,
            MathProbeOperation::SweepAdvance,
            MathProbeOperation::SweepNormalize,
            MathProbeOperation::Cancellation,
            MathProbeOperation::HalfwayRounding,
            MathProbeOperation::Overflow,
            MathProbeOperation::Underflow,
            MathProbeOperation::FmaWitness,
        ] {
            assert!(results.iter().any(|result| result.operation() == operation));
        }
    }

    #[test]
    fn native_math_probe_witness_bits_are_exact() {
        // Arrange
        let results = results();
        let expected = [
            ("cancellation", vec![0x4048_f5c3, 0x0000_0000]),
            ("halfway-rounding", vec![0x3f80_0000, 0x3f80_0002]),
            ("overflow", vec![0x7f80_0000]),
            ("underflow", vec![0x0020_0000]),
            ("fma-witness", vec![0x0000_0000]),
        ];

        // Act and Assert
        for (case_id, expected_bits) in expected {
            let result = results
                .iter()
                .find(|result| result.case_id() == case_id)
                .expect("named witness should exist");
            assert_eq!(
                result
                    .values()
                    .iter()
                    .map(|value| value.bits().bits())
                    .collect::<Vec<_>>(),
                expected_bits
            );
        }
    }
}
